### DISCUSSION TITLE
Different approach to sourcing foreign environment files

### DISCUSSION THREAD

**Discussion by dilr** [nushell/nushell#13318] [Show and tell]
This is half a show and tell, and half me asking if I can add a different approach to [sourcing foreign environments](https://www.nushell.sh/cookbook/foreign_shell_scripts.html#capturing-the-environment-from-a-foreign-shell-script) to the nushell cookbook.

My approach is a template for a polyglot script (meaning it is both a valid bash script and a valid nushell script) which first runs as bash to interpret the ".env" file, and then re executes itself as nushell, which gives you access to the environment. There is also a guard such that if the user executes it as a nushell script, the script will re-execute itself as a bash script to set up the environment before executing as a nushell script again.

The advantages of this approach is that it doesn't require a plugin like "bash-env" does, so it doesn't need the user to install anything besides nushell and bash. Also, it does not require nushell to parse the environment like "capture-forign-env" does. Therefore it is not tripped up by environmental variables which have newlines and/or quotes in them. The disadvantage is that the environmental setup needs to happen at the beginning of the script. Also, most people aren't that familiar with polyglot scripts, so it might cause some headaches if someone wanted to modify the approach.

Here is the template, with comments explaining the inner workings:
```nu-script
#!/bin/bash
# This is both a valid bash script and a valid nushell script.
# If invoked as a bash script, it sets up the
# environment using the .env file, then executes nushell.
# If invoked as a nushell script by the user,
# then it involves itself as a bash script.

# nushell interprets this as a closure,
# and does not evaluate it
# bash interprets this as a block, and so does evaluate it
{
  # bash sets up our environment here. In bash "." is the same as "source".
  # We use "." here to avoid conflicting with nushell's "source" keyword
  # USE YOUR ENVIRONMENT FILE HERE
  . ./.env

  # run nushell with the modified environment
  __RAN_DOT_ENV=true exec nu "$0" "$@" 

  # bash never reaches this line or beyond because "exec" was called
}
# the closure is over, so nushell picks up from here
# and executes to the end of the file

# This command reruns the script with bash if the environment
# was not set up before nushell was run.
def --env setup-env [
  ...argv: string # argv in case we need to rerun as bash
  ] {
  if $env.__RAN_DOT_ENV? != "true" {
    exec bash $env.CURRENT_FILE ...$argv
  }
  # unset check variable in case we run recursively
  hide-env __RAN_DOT_ENV
}

def main [...argv: string] {
  setup-env ...$argv

  # EXECUTE YOUR NUSHELL SCRIPT HERE
  print $env
}
```

I would appreciate feedback on if this approach is acceptable for the cookbook, and also feedback on if some part seems unnecessarily unclear.

EDIT: made polyglot switch significantly less clever and more  robust.

**Comment by tesujimath** [comment]
Polyglot is very cunning, I like that! 

I'm the author of what was formerly the bash-env plugin for Nushell. It was rather a pain maintaining it as a plugin, so it got a lot simpler, and became simply a [Nu module](https://github.com/tesujimath/bash-env-nushell) and accompanying [Bash script](https://github.com/tesujimath/bash-env-json).

Good to see ongoing innovation here.
